wallet/chainntfns.go

```diff
@@ -329,3 +344,151 @@ func (w *Wallet) addRelevantTx(dbtx walletdb.ReadWriteTx, rec *wtxmgr.TxRecord,
 
 	return nil
 }
+
+// birthdaySanityCheck is a helper function that ensures our birthday block
+// correctly reflects the birthday timestamp within a reasonable timestamp
+// delta. It will be run after the wallet establishes its connection with the
+// backend, but before it begins syncing.  This is done as the second part to
+// the wallet's address manager migration where we populate the birthday block
+// to ensure we do not miss any relevant events throughout rescans.
+func (w *Wallet) birthdaySanityCheck() error {
+	// We'll start by acquiring our chain backend client as we'll be
+	// querying it for blocks.
+	chainClient, err := w.requireChainClient()
+	if err != nil {
+		return err
+	}
+
+	// We'll then fetch our wallet's birthday timestamp and block.
+	birthdayTimestamp := w.Manager.Birthday()
+	var birthdayBlock waddrmgr.BlockStamp
+	err = walletdb.View(w.db, func(tx walletdb.ReadTx) error {
+		var err error
+		ns := tx.ReadBucket(waddrmgrNamespaceKey)
+		birthdayBlock, err = w.Manager.BirthdayBlock(ns)
+		return err
+	})
+
+	switch {
+	// If our wallet's birthday block has not been set yet, then this is our
+	// initial sync, so we'll defer setting it until then.
+	case waddrmgr.IsError(err, waddrmgr.ErrBirthdayBlockNotSet):
+		return nil
+
+	// Otherwise, we'll return the error if there was one.
+	case err != nil:
+		return err
+	}
+
+	log.Debugf("Starting sanity check for the wallet's birthday block "+
+		"from: height=%d, hash=%v", birthdayBlock.Height,
+		birthdayBlock.Hash)
+
+	// Now, we'll need to determine if our block correctly reflects our
+	// timestamp. To do so, we'll fetch the block header and check its
+	// timestamp.
+	candidate := birthdayBlock
+	header, err := chainClient.GetBlockHeader(&candidate.Hash)
+	if err != nil {
+		return fmt.Errorf("unable to get header for block hash %v: %v",
+			candidate.Hash, err)
+	}
+
+	// We'll go back a day worth of blocks in the chain until we find a
```

Comment:
I think this would be a very useful utility method! `firstBlockFromTimestamp(t time.Time) BlockStamp`.

This is something we want for the initial wallet rescan as well, as we also find the block there (currently by scanning from genesis).

Reply:
I agree this would be useful, but perhaps might be out of scope considering the birthday block is now being set within `syncWithChain`, which we're planning to refactor in the future anyway.

Reply:
Shouldn't be much work extracting into a method though?

Reply:
Let's opt for a minimal diff in the PR, when we have a use for the method elsewhere in the code base it can be extracted.

Reply:
They're also slightly different FWIW. At the moment, `syncWithChain` will find the first block with a timestamp greater than our birthday timestamp, while the sanity check is more just about ensuring that we have a sensible value.